toucan_connectors/snowflake/snowflake_connector.py

```diff
@@ -114,12 +117,51 @@ def get_connection_params(self):
             res['password'] = self.password.get_secret_value()
 
         if self.authentication_method == AuthenticationMethod.OAUTH:
-            res['token'] = Template(self.oauth_token).render()
+            if self.oauth_token is not None:
+                res['token'] = Template(self.oauth_token).render()
+
+            if self.oauth_args and 'access_token' in self.oauth_args:
+                res['token'] = self.oauth_args['access_token']
 
         return res
 
+    def _refresh_oauth_token(self):
+        """Regenerates an oauth token if configuration was provided and if the given token has expired."""
+        if 'token_endpoint' in self.oauth_args and 'refresh_token' in self.oauth_args:
+            if datetime.fromtimestamp(self.oauth_args['exp']) < datetime.now():
+                content_type = 'application/json'
+                # Content-Type may need to be overridden
+                if 'content_type' in self.oauth_args:
+                    content_type = self.oauth_args['content_type']
+
+                res = requests.post(
+                    Template(self.oauth_args['token_endpoint']).render(),
+                    data={
+                        'grant_type': 'refresh_token',
+                        'client_id': Template(self.oauth_args['client_id']).render(),
+                        'client_secret': Template(self.oauth_args['client_secret']).render(),
+                        'refresh_token': Template(self.oauth_args['refresh_token']).render(),
+                    },
+                    headers={'Content-Type': content_type},
+                )
+
+                if res.status_code is not requests.codes.ok:
+                    raise Exception(
+                        f'Refresh token request failed with status code: {res.status_code}'
+                    )
+
+                recieved_payload = res.json()
+
+                self.oauth_args['access_token'] = recieved_payload['access_token']
+                self.oauth_args['id_token'] = recieved_payload['id_token']
+                self.oauth_args['exp'] = datetime.now().timestamp() + recieved_payload['expires_in']
+
     def connect(self, **kwargs) -> snowflake.connector.SnowflakeConnection:
-        return snowflake.connector.connect(**self.get_connection_params(), **kwargs)
+        if self.oauth_args and self.authentication_method:
```

Comment:
ça ne serait pas plutôt `and self.authentication_method == AuthenticationMethod.OAUTH` ?

Reply:
Yes, c'est fixé.